fabric-object-builder-api-v1/src/main/java/net/fabricmc/fabric/mixin/object/builder/TradeOffersMixin.java

```diff
@@ -0,0 +1,48 @@
+/*
+ * Copyright (c) 2016, 2017, 2018, 2019 FabricMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.mixin.object.builder;
+
+import com.google.common.collect.ImmutableMap;
+import org.spongepowered.asm.mixin.Mixin;
+import org.spongepowered.asm.mixin.injection.At;
+import org.spongepowered.asm.mixin.injection.Redirect;
+import org.spongepowered.asm.mixin.injection.Slice;
+
+import net.minecraft.item.Item;
+import net.minecraft.util.registry.Registry;
+import net.minecraft.village.TradeOffers;
+import net.minecraft.village.VillagerType;
+
+import net.fabricmc.fabric.impl.object.builder.FabricVillagerType;
+
+@Mixin(TradeOffers.class)
+public abstract class TradeOffersMixin {
+	/**
+	 * @reason Due to how Minecraft handles type aware trades, the game will throw an exception and crash if a villager type does not have an entry in the type aware trades.
+	 * We attempt to fix this issue by adding our own villager types into the map before it is built.
+	 * However this requires that all villager types are registered before any trade offers are setup.
+	 */
+	@Redirect(method = "method_16929", at = @At(value = "INVOKE", target = "Lcom/google/common/collect/ImmutableMap$Builder;build()Lcom/google/common/collect/ImmutableMap;"), slice = @Slice(from = @At(value = "FIELD", target = "Lnet/minecraft/item/Items;PUFFERFISH:Lnet/minecraft/item/Item;"), to = @At(value = "INVOKE", shift = At.Shift.AFTER, target = "Lcom/google/common/collect/ImmutableMap$Builder;build()Lcom/google/common/collect/ImmutableMap;")))
```

Comment:
Imo we should just make the map non-immutable if possible

Reply:
Making the map mutable solves nothing.

So we make the map mutable. Okay how do we handle mods adding custom villager types that are not present on the map? Below is the code that will crash the game if you type isn't on the map.
```java
Registry.VILLAGER_TYPE.stream().filter((villagerType) -> {
    return !map.containsKey(villagerType);
}).findAny().ifPresent((villagerType) -> {
    throw new IllegalStateException("Missing trade for villager type: " + Registry.VILLAGER_TYPE.getId(villagerType));
});
```

Reply:
Making the map mutable at least makes it not dependent on load order. As a result, you don't need to create `FabricVillagerType` class, but just call a `put` in that map. Also you can just consider patch that `stream()` call to return an empty stream.

Reply:
The map is private, so some util method would be needed since AWs would be needed for an accessor. And for someone to add their own entry, they need to traverse the `Map<VillagerProfession, Int2ObjectMap<TradeOffers.Factory[]>>` if we supply no mechanism for knowing when the trade is created.